app/code/Magento/InventoryShipping/Observer/VirtualSourceDeductionProcessor.php

```diff
@@ -0,0 +1,271 @@
+        GetSourceItemBySourceCodeAndSku $getSourceItemBySourceCodeAndSku,
+        GetSkusByProductIdsInterface $getSkusByProductIds,
+        GetStockItemConfigurationInterface $getStockItemConfiguration,
+        DefaultSourceProviderInterface $defaultSourceProvider,
+        Json $jsonSerializer,
+        InventoryRequestInterfaceFactory $inventoryRequestFactory,
+        ItemRequestInterfaceFactory $itemRequestFactory,
+        SourceSelectionServiceInterface $sourceSelectionService,
+        GetDefaultSourceSelectionAlgorithmCodeInterface $getDefaultSourceSelectionAlgorithmCode
+    ) {
+        $this->reservationBuilder = $reservationBuilder;
+        $this->appendReservations = $appendReservations;
+        $this->sourceItemsSave = $sourceItemsSave;
+        $this->stockByWebsiteIdResolver = $stockByWebsiteIdResolver;
+        $this->getSourceItemBySourceCodeAndSku = $getSourceItemBySourceCodeAndSku;
+        $this->getSkusByProductIds = $getSkusByProductIds;
+        $this->getStockItemConfiguration = $getStockItemConfiguration;
+        $this->defaultSourceProvider = $defaultSourceProvider;
+        $this->jsonSerializer = $jsonSerializer;
+        $this->inventoryRequestFactory = $inventoryRequestFactory;
+        $this->itemRequestFactory = $itemRequestFactory;
+        $this->sourceSelectionService = $sourceSelectionService;
+        $this->getDefaultSourceSelectionAlgorithmCode = $getDefaultSourceSelectionAlgorithmCode;
+    }
+
+    /**
+     * @param EventObserver $observer
+     * @return void
+     * @throws LocalizedException
+     */
+    public function execute(EventObserver $observer)
+    {
+        /** @var \Magento\Sales\Model\Order\Invoice $invoice */
+        $invoice = $observer->getEvent()->getInvoice();
+
+        if ($invoice->getOrigData('entity_id')) {
+            return;
+        }
+
+        $selectionRequestItems = [];
+        foreach ($invoice->getItems() as $invoiceItem) {
+            if (!$this->isValidItem($invoiceItem)) {
+                continue;
+            }
+
+            $itemSku = $invoiceItem->getSku() ?: $this->getSkusByProductIds->execute(
+                [$invoiceItem->getProductId()]
+            )[$invoiceItem->getProductId()];
+            $qty = $this->castQty($invoiceItem->getOrderItem(), $invoiceItem->getQty());
+
+            $selectionRequestItems[] = $this->itemRequestFactory->create([
+                'sku' => $itemSku,
+                'qty' => $qty,
+            ]);
+        }
+
+        if (empty($selectionRequestItems)) {
+            return;
+        }
+
+        $order = $invoice->getOrder();
+        $websiteId = $order->getStore()->getWebsiteId();
+        $stockId = (int)$this->stockByWebsiteIdResolver->get((int)$websiteId)->getStockId();
+
+        /** @var InventoryRequestInterface $inventoryRequest */
+        $inventoryRequest = $this->inventoryRequestFactory->create([
+            'stockId' => $stockId,
+            'items' => $selectionRequestItems
+        ]);
```

Comment:
this logic
```
   $selectionRequestItems = [];
        foreach ($invoice->getItems() as $invoiceItem) {
            if (!$this->isValidItem($invoiceItem)) {
                continue;
            }

            $itemSku = $invoiceItem->getSku() ?: $this->getSkusByProductIds->execute(
                [$invoiceItem->getProductId()]
            )[$invoiceItem->getProductId()];
            $qty = $this->castQty($invoiceItem->getOrderItem(), $invoiceItem->getQty());

            $selectionRequestItems[] = $this->itemRequestFactory->create([
                'sku' => $itemSku,
                'qty' => $qty,
            ]);
        }

        if (empty($selectionRequestItems)) {
            return;
        }

        $order = $invoice->getOrder();
        $websiteId = $order->getStore()->getWebsiteId();
        $stockId = (int)$this->stockByWebsiteIdResolver->get((int)$websiteId)->getStockId();

        /** @var InventoryRequestInterface $inventoryRequest */
        $inventoryRequest = $this->inventoryRequestFactory->create([
            'stockId' => $stockId,
            'items' => $selectionRequestItems
        ]);
        $sourceSelectionResult = $this->sourceSelectionService->execute(
            $inventoryRequest,
            $this->getDefaultSourceSelectionAlgorithmCode->execute()
        );
```

could be also extracted to dedicated service source selection and placed into dedicated module `InventorySourceSleection`

Reply:
I introduced `\Magento\InventoryShipping\Model\InventoryRequestFromInvoiceFactory` which works like `\Magento\InventorySourceSelection\Model\InventoryRequestFromOrderFactory`. Since `InventorySourceSelection` and `InventorySourceDeduction` should not know anything about order and invoice I think this factory should not be part of `InventorySourceSelection`... Probably we need to move `\Magento\InventorySourceSelection\Model\InventoryRequestFromOrderFactory` some where else...